From 566a7bb8e21825c841d0a24fb0786ba7202426fb Mon Sep 17 00:00:00 2001
From: Hong-Mei Li <a21834@motorola.com>
Date: Wed, 22 Apr 2015 17:55:01 -0700
Subject: [PATCH 04/23] libc: arm64: memcpy optimization

Optimize memcpy to prefetch several cache lines.
We can achieve about 300 memory quadrant score improvement with this patch.

Change-Id: Ie7ed07c22ee241d4e41da360eba67535efb707c8
Signed-off-by: Hong-Mei Li <a21834@motorola.com>
Reviewed-on: http://gerrit.mot.com/740769
SME-Granted: SME Approvals Granted
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Zhi-Ming Yuan <a14194@motorola.com>
Reviewed-by: Xin Guan <a18772@motorola.com>
SLTApproved: Xin Guan <a18772@motorola.com>
Reviewed-by: Yi-Wei Zhao <gbjc64@motorola.com>
Reviewed-by: Jing Ji <a5705c@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
Reviewed-on: http://gerrit.mot.com/765184
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Signed-off-by: mydongistiny <jaysonedson@gmail.com>
---
 libc/arch-arm64/generic/bionic/memcpy_base.S | 1 +
 1 file changed, 1 insertion(+)

diff --git a/libc/arch-arm64/generic/bionic/memcpy_base.S b/libc/arch-arm64/generic/bionic/memcpy_base.S
index f85062408..892c082a9 100644
--- a/libc/arch-arm64/generic/bionic/memcpy_base.S
+++ b/libc/arch-arm64/generic/bionic/memcpy_base.S
@@ -184,6 +184,7 @@ L(copy_long):
 	ldp	B_l, B_h, [src, 32]
 	ldp	C_l, C_h, [src, 48]
 	ldp	D_l, D_h, [src, 64]!
+       prfm    pldl1strm, [src, #(4*64)]
 	subs	count, count, 128 + 16	/* Test and readjust count.  */
 	b.ls	2f
 1:
-- 
2.11.0

